File: pfe-java/service/src/main/java/models/AuctionRooms.java
package models;

import akka.actor.ActorRef;
import akka.actor.Props;
import akka.util.Timeout;
import play.libs.Akka;
import play.libs.F;
import scala.concurrent.Future;
import scala.concurrent.duration.Duration;

import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;

import static akka.pattern.Patterns.ask;

/* In-memory publish/subscribe system */
public class AuctionRooms {

    final ActorRef ref = Akka.system().actorOf(Props.create(AuctionRoomsActor.class));
    final Timeout t = new Timeout(Duration.create(1, TimeUnit.SECONDS));

    /**
     * Subscribe to bid notifications for a given item
     * @param id Item id
     * @param subscriber Notification callback
     * @return A pair containing the current bids and a Subscription object (for un-subscribing)
     */
    public F.Promise<F.Tuple<List<Bid>, Subscription>> subscribe(Long id, Consumer<Bid> subscriber) {
        return F.Promise.wrap((Future)ask(ref, new AuctionRoomsActor.Subscribe(id, subscriber), t));
    }

    /**
     * Bid for an item
     * @param id Item id
     * @param name User name
     * @param price Bid price
     */
    public void bid(Long id, String name, Double price) {
        ref.tell(new AuctionRoomsActor.ItemBid(id, name, price), null);
    }

    public static class Subscription {
        private final Long id;
        private final Consumer<Bid> subscriber;
        private final ActorRef ref;

        public Subscription(ActorRef ref, Long id, Consumer<Bid> subscriber) {
            this.ref = ref;
            this.id = id;
            this.subscriber = subscriber;
        }

        /**
         * Tell the subscription system that this subscription is not anymore interested in receiving notifications
         */
        public void cancel() {
            ref.tell(new AuctionRoomsActor.Unsubscribe(id, subscriber), null);
        }
    }

    public static class Bid {
        public String name;
        public Double price;

        public Bid() {}

        public Bid(String name, Double price) {
            this.name = name;
            this.price = price;
        }
    }

}

File: pfe-scala/service/src/main/scala/models/AuctionRooms.scala
package models

import akka.actor.{Actor, ActorSystem, Props}
import akka.stream.{ActorMaterializer, Materializer, OverflowStrategy}
import akka.stream.scaladsl.{Sink, Source}

import scala.concurrent.Future

class AuctionRoomsActor extends Actor {
  import AuctionRoomsActor._

  implicit val materializer: Materializer = ActorMaterializer()
  var rooms = Map.empty[Long, Room]

  class Room {
    var bids = Map.empty[String, Double]
    val (ref, notifications) =
      Source.actorRef(1000, OverflowStrategy.dropNew)
        .toMat(Sink.asPublisher[(String, Double)](fanout = true)) { case (ref, publisher) => (ref, Source.fromPublisher(publisher)) }
        .run()

    def stateAndNotifications() = (bids, notifications)

    def addBid(name: String, price: Double): Unit = {
      if (bids.forall { case (_, p) => p < price}) {
        bids += name -> price
        ref ! (name -> price)
      }
    }

  }

  def lookupOrCreate(id: Long): Room = rooms.getOrElse(id, {
    val room = new Room
    rooms += id -> room
    room
  })

  def receive = {
    case Notifications(id) =>
      sender() ! lookupOrCreate(id).stateAndNotifications()
    case ItemBid(id, name, price) =>
      lookupOrCreate(id).addBid(name, price)
  }

}

object AuctionRoomsActor {
  case class Notifications(id: Long)
  case class ItemBid(id: Long, name: String, price: Double)
}

class AuctionRooms(actorSystem: ActorSystem) {

  import AuctionRoomsActor._
  import akka.pattern.ask

  import concurrent.duration.DurationInt

  implicit val timeout: akka.util.Timeout = 1.second

  private lazy val ref = actorSystem.actorOf(Props(new AuctionRoomsActor))

  /**
   * Ask for the notifications stream of an auction room
   * @param id Item id
   */
  def notifications(id: Long): Future[(Map[String, Double], Source[(String, Double), _])] =
    (ref ? Notifications(id)).mapTo[(Map[String, Double], Source[(String, Double), _])]

  /**
   * Bid for an item
   * @param id Item id
   * @param name User name
   * @param price Bid price
   */
  def bid(id: Long, name: String, price: Double): Unit =
    ref ! ItemBid(id, name, price)

}